### ISSUE TITLE
namespace not found

### ISSUE THREAD

**Issue by sebolabs** [argoproj/argo-cd#11625]
**Describe the bug**
I'm getting a `namespace not found` error when deploying an app.
It worked yesterday but stopped working today. Since yesterday all nodes have only been scaled down and up again. No configuration was changed.

App definition:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  labels:
    argocd.argoproj.io/instance: test-hello-world-app
  name: test-hello-world
  namespace: test-hello-world
spec:
  destination:
    namespace: test-hello-world
    server: 'https://kubernetes.default.svc'
  project: sebolabs-test
  source:
    helm:
      valueFiles:
        - values.yaml
        - values-test.yaml
    path: apps/hello-world
    repoURL: 'https://...'
    targetRevision: trying-out-things
  syncPolicy:
    automated:
      prune: true
    retry:
      backoff:
        duration: 5s
        factor: 2
        maxDuration: 1m
      limit: 1
    syncOptions:
      - CreateNamespace=true
```

The `sebolabs-test` project accepts `test-*` namespaces + ArgoCD is configured with `application.namespaces: "test-*"`.
Just like it was explained yesterday [here](https://github.com/argoproj/argo-cd/issues/11597).

As I was running out of ideas on how to get it working again I updated ArgoCD to v2.5.4 but that didn't change anything.

**To Reproduce**

See above.

**Expected behavior**

The `test-hello-world` should be automatically created and the app should be successfully deployed.

**Screenshots**

N/A

**Version**

```shell
v2.5.4+86b2dde
```

**Logs**

```
time="2022-12-08T21:29:29Z" level=info msg=Syncing application=argocd/test-hello-world-app skipHooks=false started=false syncId=00008-QWOTG
time="2022-12-08T21:29:29Z" level=info msg="Tasks (dry-run)" application=argocd/test-hello-world-app syncId=00008-QWOTG tasks="[Sync/0 resource argoproj.io/Application:test-hello-world/test-hello-world nil->obj (,,)]"
time="2022-12-08T21:29:29Z" level=info msg="Applying resource Application/test-hello-world in cluster: https://172.20.0.1:443, namespace: test-hello-world"
time="2022-12-08T21:29:29Z" level=info msg="Updating operation state. phase: Running -> Running, message: 'one or more objects failed to apply, reason: namespaces \"test-hello-world\" not found. Retrying attempt #4 at 9:28PM.' -> 'one or more tasks are running'" application=argocd/test-hello-world-app syncId=00008-QWOTG
time="2022-12-08T21:29:29Z" level=info msg="Applying resource Application/test-hello-world in cluster: https://172.20.0.1:443, namespace: test-hello-world"
time="2022-12-08T21:29:29Z" level=info msg="Apply failed" application=argocd/test-hello-world-app dryRun=false message="namespaces \"test-hello-world\" not found" syncId=00008-QWOTG task="Sync/0 resource argoproj.io/Application:test-hello-world/test-hello-world nil->obj (,,)"
time="2022-12-08T21:29:29Z" level=info msg="Adding resource result, status: 'SyncFailed', phase: 'Failed', message: 'namespaces \"test-hello-world\" not found'" application=argocd/test-hello-world-app kind=Application name=test-hello-world namespace=test-hello-world phase=Sync syncId=00008-QWOTG
time="2022-12-08T21:29:29Z" level=info msg="Updating operation state. phase: Running -> Failed, message: 'one or more tasks are running' -> 'one or more objects failed to apply, reason: namespaces \"test-hello-world\" not found'" application=argocd/test-hello-world-app syncId=00008-QWOTG
```

**Comment by fengshunli** [comment]
<img width="539" alt="image" src="https://user-images.githubusercontent.com/18587688/206965499-86531ffb-99d2-4361-b3ee-8539a6f6ab4f.png">
.spec.syncPolicy.syncOption can be solved by adding PruneLast=true parameter

**Comment by sebolabs** [comment]
Unfortunately, that didn't work and I think it could only be considered (by design) if the app got created.
It keeps on saying that `... one or more objects failed to apply, reason: namespaces "test-hello-world" not found. Retrying attempt ...`

**Comment by crenshaw-dev** [comment]
How are you deploying the `test-hello-world` app? From an app-of-apps, applicationset, kubectl or other?

**Comment by crenshaw-dev** [comment]
Looking at the logs, I think `argocd/test-hello-world-app` is your parent app. That's the one that needs `CreateNamespace=true`.

**Comment by sebolabs** [comment]
That's correct, the `argocd/test-hello-world-app` is the parrent app and here's its manifest:
```
project: sebolabs-test
source:
  repoURL: 'https://...'
  path: hello-world
  targetRevision: testing
destination:
  server: 'https://kubernetes.default.svc'
syncPolicy:
  automated:
    prune: true
    selfHeal: true
  syncOptions:
    - Validate=false
    - CreateNamespace=true
    - PrunePropagationPolicy=foreground
    - PruneLast=true
    - RespectIgnoreDifferences=true
  retry:
    limit: 10
    backoff:
      duration: 10s
      factor: 2
      maxDuration: 3m
```

```
$ kubectl get apps -A | grep hello-world
argocd      test-hello-world-app           OutOfSync     Healthy
```

**Comment by crenshaw-dev** [comment]
If you sync that app, it should create the namespace for the child app.

**Comment by sebolabs** [comment]
The thing is that I can't, the parent (top-level) app will be in sync if child items get synced. At least that's how I believe it works.
Here's what I see when I click Sync for the parent app:
<img width="1167" alt="image" src="https://user-images.githubusercontent.com/8235212/207433362-f5883f29-7079-47df-a6e2-e156b3b1536b.png">

**Comment by crenshaw-dev** [comment]
> the parent (top-level) app will be in sync if child items get synced

Isn't that okay? Why wouldn't you want the parent app to be in sync?

**Comment by crenshaw-dev** [comment]
I think the order needs to go "sync parent app" (now child app exists, so you can...) "sync child app(s)".

**Comment by sebolabs** [comment]
I would, but you can see on the screenshot what's there to be synchronized when I hit Sync for the parent app:
The only resource is the child app (`test-hello-world/test-hello-world`), this is why I said **I believe** that the parent app can show up as in sync if the child app gets synced.
And the latter won't happen because the namespace is not there even though the create namespace sync option is enabled all over the place.
I don't get that because it worked and stopped after I deleted everything and tried to let it be recreated, just like I explained in the beginning.

**Comment by sebolabs** [comment]
See this, example:
![image](https://user-images.githubusercontent.com/8235212/207438036-fc4b54fb-5533-464a-85ea-434ae7563d83.png)
I clicked Sync **individually** for both webhooks there and when they got synced the `aws-load-balancer-controller` app automatically appeared as Synced.

**Comment by crenshaw-dev** [comment]
> the latter won't happen because the namespace is not there even though the create namespace sync option is enabled all over the place

Gotcha. The parent app's sync CreateNamespace option should be all that matters. The _order_ of syncing matters, because until the parent app is synced there is 1) probably no namespace called `test-hello-world` and 2) no child app to sync.

What happens if you try to sync the parent app? Do you get the error message that is in the original post?

**Comment by crenshaw-dev** [comment]
Apologies for the back-and-forth. Understanding the app-of-apps setup was the part I was missing, even though I think all the info was in the original post. :-)

**Comment by sebolabs** [comment]
So again :) when I hit Sync for the parent app it is trying to create the child app and complains that the namespace is not there.
<img width="380" alt="image" src="https://user-images.githubusercontent.com/8235212/207440933-169591db-9574-4742-888b-374018a40d54.png">

**Comment by sebolabs** [comment]
No worries. For testing, and as expected, I created the namespace with kubectl and everything got nicely synced. Then I went and deleted both the app and the namespace, redeployed the parent app, and ended up with the same error.

**Comment by crenshaw-dev** [comment]
Okee-doke, here's a minimal reproduction of the issue:

app.yaml:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-hello-world-app
  namespace: argocd
spec:
  project: default
  source:
    repoURL: https://github.com/crenshaw-dev/argo-cd-11625
    path: resource
  destination:
    server: 'https://kubernetes.default.svc'
    # uncomment this to demo the bug
    #namespace: test-hello-world
  syncPolicy:
    syncOptions:
      - CreateNamespace=true
```

contents of resource to be deployed:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: test-hello-world
```

This conflicts with my understanding of how the `CreateNamespace` is supposed to work. 

My belief was that it would create a namespace each time it encountered a resource destined for a namespace which does not exist.

The reality is that it just creates the namespace specified in `spec.destination.namespace`, regardless of whether any resource actually _uses_ that namespace.

@ashutosh16, @leoluz was I just incorrect about how this feature works? Or is this a regression?

**Comment by crenshaw-dev** [comment]
The behavior is the same for 2.4.17 and 2.3.11.

**Comment by crenshaw-dev** [comment]
Issue was previously identified here: https://github.com/argoproj/argo-cd/issues/6997

**Comment by crenshaw-dev** [comment]
I shudder to say it, but I think we need `CreateNamespaces=true`.

**Comment by leoluz** [comment]
> My belief was that it would create a namespace each time it encountered a resource destined for a namespace which does not exist.
The reality is that it just creates the namespace specified in spec.destination.namespace, regardless of whether any resource actually uses that namespace.@ashutosh16, @leoluz was I just incorrect about how this feature works? Or is this a regression?

That is correct. Clarified it in the docs to avoid this type of confusion in the future: https://github.com/argoproj/argo-cd/pull/11723

@sebolabs @crenshaw-dev can we close this issue as the behaviour is clarified?

**Comment by crenshaw-dev** [comment]
@sebolabs if `spec.destination.namespace` doesn't solve your use case (i.e. you need to deploy to multiple or not-yet-known namespaces), would you be willing to open an enhancement request?

tbh, I think creating the _resources_ namespaces should be the default behavior, but we'd need to add the feature on a new SyncOption to preserve backwards compatibility.

**Comment by nicolas-vivot** [comment]
The ApplicationSet example for the git-generator-directory is a good example to reproduce this bug

https://github.com/argoproj/argo-cd/tree/master/applicationset/examples/git-generator-directory

You just have to add the namespace restrictions to the project name used in the example.

**Comment by bkk-bcd** [comment]
Did this get fixed? I'm seeing the same issue with v2.7.3

**Comment by bkk-bcd** [comment]
Nevermind, I forgot the obvious:

```
      syncPolicy:
        automated: {}
        syncOptions:
          - CreateNamespace=true
 ```

**Comment by nicl-dev** [comment]
We are still seeing the same error with
 ``` 
syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```

**Comment by mrclrchtr** [comment]
Sam problem here.. I cannot find out, whats wrong..

**Comment by cioban** [comment]
I'm facing the same problem here.

**Comment by zimmertr** [comment]
Also seeing it with the latest version of ArgoCD and K8s when targeting this `kustomization.yml` file with an ApplicationSet using the `git` generator with `spec.template.spec.syncPolicy.syncOptions["CreateNamespace=true"]` set. 

```yaml
# https://www.talos.dev/v1.5/kubernetes-guides/configuration/deploy-metrics-server/

resources:
  - https://raw.githubusercontent.com/alex1989hu/kubelet-serving-cert-approver/main/deploy/standalone-install.yaml

helmCharts:
  - name: metrics-server
    repo: https://kubernetes-sigs.github.io/metrics-server/
    version: 3.11.0 # https://artifacthub.io/packages/helm/metrics-server/metrics-server?modal=changelog
    releaseName: metrics-server
    namespace: metrics-server
    valuesFile: values.yml
    includeCRDs: true
```

**Comment by zimmertr** [comment]
Removing the leading spaces for the `CreateNamespace=true` sync option solved the issue as per [this comment](https://cloud-native.slack.com/archives/C01TSERG0KZ/p1641403770264600?thread_ts=1641403189.264100&cid=C01TSERG0KZ) on the CNCF Slack. 

```
         syncOptions:
-          - CreateNamespace=true
+        - CreateNamespace=true
```

**Comment by Honken77** [comment]
Just had this issue when trying to recreate an app. I had to remove the namespace to purge some annotations we set with `managedNamespaceMetadata`. We are running v2.6.14, though.

It's an app-of-apps structure where I removed the whole app from the mother-app but it was still unable to recreate the namespace. It could only create the cluster resources that weren't namespace-bound.

**Comment by hahawu42** [comment]
Hi @sebolabs, given your original app definition above it can be that you try to deploy your argocd application into the wrong namespace (test-hello-world), because your `metadata.namespace` and your `spec.destination.namespace` are identical. Is your ArgoCD application really deployed into "test-hello-world" namespace? Otherwise you need to change the `metadata.namespace` to the namespace where you deployed ArgoCD itself.

**Comment by nicl-dev** [comment]
We just added a new application to our app-of-apps deplyoment using kustomize and ran into this error - again. We are using the following definition:

```
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: dev-argoworkflows
  namespace: argoworkflows
spec:
  project: ourproject
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
  destination:
    namespace: argoworkflows
    server: https://kubernetes.default.svc
  source:
    repoURL: https://ourgitlab.com/ourproject/deployment.git
    targetRevision: HEAD
    path: LIVE/dev/infra/argoworkflows
```

We do have a kustomization.yaml that points to our Argo-CD application which uses another kustomization.yaml that points to our applications, including this new deployment.

**Comment by didlawowo** [comment]
have the same issue, with classic app pattern (not app of apps)

create namespace doesn't work...

**Comment by bert-jan** [comment]
I also have the same issue (with and without) apps-of-apps in version v2.8.6.

**Comment by caiocsgomes** [comment]
Same here with the latest version

**Comment by houranii** [comment]
It worked for me when specifying 
    syncOptions:
      - CreateNamespace=true
      - PruneLast=true 
    PruneLast=true
    
    Using latest helm
    
    Ref: 
    https://github.com/argoproj/argo-cd/issues/4438#issuecomment-1213474821

**Comment by GenPage** [comment]
I'm also on latest version and having this issue.

I have `spec.destination.namespace` set with the proper `CreateNamespace=true` syncOption.

This works if I terminate the initial auto sync, manually trigger the sync via the UI and select `AutoCreateNamespace`.

**Comment by yiwwhl** [comment]
> I'm also on latest version and having this issue.
> 
> I have `spec.destination.namespace` set with the proper `CreateNamespace=true` syncOption.
> 
> This works if I terminate the initial auto sync, manually trigger the sync via the UI and select `AutoCreateNamespace`.

Hello there! I've encountered an issue with Kubernetes that I was hoping you might be able to help me with.
I'm experiencing the same problem you mentioned earlier. When I try to sync, I get an error saying it can't find the namespace. My Kubernetes version is 1.23.12. I'm wondering if this issue might be related to the Kubernetes version.
Have you found a way to resolve this? If so, could you share how you configured your setup to fix it? Any insights or advice would be greatly appreciated!
Thank you in advance for your help!

Here's the manifest for my project:

<img width="802" alt="image" src="https://github.com/user-attachments/assets/86e835b6-9f00-4b3c-9038-6ff6def414b1">

**Comment by zswanson** [comment]
I've seen the same thing as @GenPage in 2.11 - namespace does not create when an Appset first generates the app with all the appropriate options, I have to sync manually to get it to create the ns.

**Comment by fuzolan** [comment]
I have the same issue with an app who has hooks enabled. it does not matter if i do this manually or with options 
"- CreateNamespace=true
- PruneLast=true"

**Comment by Diliz** [comment]
Same here, got the error when using the argocd cli, not the ui.

The namespace resource appear when I manually Sync from the argocd ui, and it works this way.

The namespace resource does not appear as part of the application when I try to sync it with the argocd cli.

Maybe the Sync command have an issue if not anything else, dunno why it would work in the ui and not via the CLI basically...

**Comment by Faithtosin** [comment]
The problem here is, the ArgoCD application namespace needs to already exist. So set the ArgoCD application namespace to `argocd` and then destination namespace to the new namespace. 

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  labels:
    argocd.argoproj.io/instance: test-hello-world-app
  name: test-hello-world
  namespace: argocd  #👈  this should exist
spec:
  destination:
    namespace: test-hello-world #👈  this will be created
    server: 'https://kubernetes.default.svc'
  project: sebolabs-test
  source:
    helm:
      valueFiles:
        - values.yaml
        - values-test.yaml
    path: apps/hello-world
    repoURL: 'https://...'
    targetRevision: trying-out-things
  syncPolicy:
    automated:
      prune: true
    retry:
      backoff:
        duration: 5s
        factor: 2
        maxDuration: 1m
      limit: 1
    syncOptions:
      - CreateNamespace=true
```

**Comment by reggie-k** [comment]
TLDR:

```
syncOptions:   
- CreateNamespace=true
```
only auto-creates the namespace in `spec.destination.namespace` of the Application and NOT the namespaces of the resources that the Application manages. In other words, if the namespace of one of the resources does not exist and is different from `spec.destination.namespace` of the Application, it will not be auto-created and will result in an Application sync failure.

Also, it was reported that 
```
syncOptions:
- PruneLast=true
```
is sort of a workaround, that makes the managed resource namespace be created. I couldn't reproduce this, adding PruneLast for me still results in a sync failure and in the ns not being created.

Docs were already improved to emphasize the expected sync behavior.

Action items:
- Improve the docs further
- Open a feature request to make Argo CD auto-create the namespaces of the managed resources in addition to the `spec.destination.namespace` of the Application
